L1Trigger/Phase2L1ParticleFlow/src/CaloClusterer.cc

```diff
@@ -0,0 +1,640 @@
+
+#include <cassert>
+
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include "DataFormats/Common/interface/RefToPtr.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+const float l1tpf_calo::Phase1Grid::phase1_towerEtas_[l1tpf_calo::Phase1Grid::phase1_nEta_] = {
+    0,     0.087, 0.174, 0.261, 0.348, 0.435, 0.522, 0.609, 0.696, 0.783, 0.870, 0.957, 1.044, 1.131,
+    1.218, 1.305, 1.392, 1.479, 1.566, 1.653, 1.740, 1.830, 1.930, 2.043, 2.172, 2.322, 2.5,   2.650,
+    2.853, 3.139, 3.314, 3.489, 3.664, 3.839, 4.013, 4.191, 4.363, 4.538, 4.716, 4.889, 5.191};
+const float l1tpf_calo::Phase2Grid::phase2_towerEtas_[l1tpf_calo::Phase2Grid::phase2_nEta_] = {
+    0,     0.087, 0.174, 0.261, 0.348, 0.435, 0.522, 0.609, 0.696, 0.783, 0.870, 0.957, 1.044, 1.131, 1.218, 1.305,
+    1.392, 1.479, 1.564, 1.648, 1.732, 1.817, 1.901, 1.986, 2.071, 2.155, 2.240, 2.324, 2.409, 2.493, 2.577, 2.662,
+    2.747, 2.831, 2.915, 3.0,   3.139, 3.314, 3.489, 3.664, 3.839, 4.013, 4.191, 4.363, 4.538, 4.716, 4.889, 5.191};
+
+l1tpf_calo::Phase1GridBase::Phase1GridBase(
+    int nEta, int nPhi, int ietaCoarse, int ietaVeryCoarse, const float *towerEtas)
+    : Grid(2 * ((ietaCoarse - 1) * nPhi + (ietaVeryCoarse - ietaCoarse) * (nPhi / 2) +
+                (nEta - ietaVeryCoarse + 1) * (nPhi / 4))),
+      nEta_(nEta),
+      nPhi_(nPhi),
+      ietaCoarse_(ietaCoarse),
+      ietaVeryCoarse_(ietaVeryCoarse),
+      towerEtas_(towerEtas),
+      cell_map_(2 * nEta * nPhi, -1) {
+  int icell = 0;
+  for (int ie = -nEta_; ie <= nEta_; ++ie) {
+    int absie = std::abs(ie);
+    for (int iph = 1; iph <= nPhi_; ++iph) {
+      if (!valid_ieta_iphi(ie, iph))
+        continue;
+      ieta_[icell] = ie;
+      iphi_[icell] = iph;
+      eta_[icell] = (ie > 0 ? 0.5 : -0.5) * (towerEtas_[absie - 1] + towerEtas_[absie]);
+      etaWidth_[icell] = (towerEtas_[absie] - towerEtas_[absie - 1]);
+      phiWidth_[icell] = 2 * M_PI / nPhi_;
+      if (absie >= ietaVeryCoarse_)
+        phiWidth_[icell] *= 4;
+      else if (absie >= ietaCoarse_)
+        phiWidth_[icell] *= 2;
+      phi_[icell] = (iph - 1) * 2 * M_PI / nPhi_ + 0.5 * phiWidth_[icell];
+      if (phi_[icell] > M_PI)
+        phi_[icell] -= 2 * M_PI;
+      std::fill(neighbours_[icell].begin(), neighbours_[icell].end(), -1);
+      cell_map_[(ie + nEta_) + 2 * nEta_ * (iph - 1)] = icell;
+      icell++;
+    }
+  }
+  assert(unsigned(icell) == ncells_);
+  // now link the cells
+  for (icell = 0; icell < int(ncells_); ++icell) {
+    int ie = ieta_[icell], iph = iphi_[icell];
+    int ineigh = 0;
+    for (int deta = -1; deta <= +1; ++deta) {
+      for (int dphi = -1; dphi <= +1; ++dphi) {
+        if (deta == 0 && dphi == 0)
+          continue;
+        neighbours_[icell][ineigh++] = imove(ie, iph, deta, dphi);
+      }
+    }
+  }
+  //// consistency check 1: check that find_cell works
```

Comment:
delete commented-out code

Reply:
these are useful to be uncommented in development